docs/development/secure-development-process.md

```diff
@@ -0,0 +1,204 @@
+# Secure Development Process
+
+It is important to build security into your software development process. Some of the key elements to address in the development process for Node.js modules and applications include:
+
+* Managing dependencies
+* Managing access and content of public and private data stores
+  such as npm and github 
+* Writing defensive code
+* Limiting required execution privileges
+* Supporting logging and monitoring
+* Externalizing secrets
+* Maintaining a secure and up to date foundation for deployed
+ applications
+* Maintaining individual modules
+
+Many modern applications are deployed through containers so that is the focus of this section. However, most of the guidance should apply to other packaging/deployment methods with the requirement to maintain secure base container images replaced with the requirement to keep the bare metal machines used for deployments secure and up to date.
+
+## Recommended Components
+
+N/A
+
+## Guidance
+
+### Managing dependencies
+
+Dependencies are a common vector for security vulnerabilities
+and it is important to manage depenencies carefully. This
+is covered in the
+[choosing and vetting dependencies](./dependencies.md) section of the reference architecture.
+
+### Managing access and content of public and private data stores
+
+Modern development flows often use public and private data stores
+including npm and github. We recommend the following:
+
+* Enable Two-Factor Authentication (2FA)
+* Ensure projects use a `.npmignore`, `.gitignore`, etc. to avoid 
+  accidentally publishing secrets.
+
+2FA is important to ensure the integrity of the committed code and published assets.
+
+A `.npmrc` file is often needed to do npm installs, particularly
+if you have private modules. Avoid leaking information in the `.npmrc`
+file when building containers through one of these options:
+* Use 2 stages builds and avoiding copying the `.npmrc` to the final image
+* Avoid adding the secrets to any image in the build process by mounting
+  secrets into containers during the build process. Builda has some functionality
+  build in to make this easier. The
+  [sneak secreats into your containers](https://projectatomic.io/blog/2018/06/sneak-secrets-into-containers)
+  article provides some info on how to do this.
+* Delete the `.npmrc` file from the final image AND compressing images
+  to flatten layers (least perferred).
+
+### Writing defensive code
+
+Defensive code should:
+* Avoid global state.
+* Set NODE_ENV to production.
+* use `strict` mode 
+* Include good exception handling
+* Avoid complex regexes
+* Limit the attack surface
+* Harden external endpoints
+* Take precautions against brute force and denial of service 
+  attacks
+* Avoid leaking info through errors
+* Limit data sent back to front ends
+* Discard sensitive data after use
+* Validate input and output
+
+**Avoid global state**
```

Comment:
Should these be headings (e.g. `####`) rather than bold text?

Reply:
I had tried #### but they did not show up well. That's why I switched to using bold instead.